theano/gpuarray/subtensor.py

```diff
@@ -1182,3 +1182,37 @@ def infer_shape(self, node, shapes):
             diag_size = T.minimum(dim1, dim2)
         out_shape.append(diag_size)
         return [tuple(out_shape)]
+
+
+class GpuAllocDiag(Subtensor):
+    __props__ = ("offset",)
+
+    def __init__(self, offset=0, axis1=0, axis2=1, view=False):
+        self.offset = offset
+
+    def make_node(self, _x):
+        ctx_name = infer_context_name(_x)
+        x = as_gpuarray_variable(_x, ctx_name)
+
+        if x.ndim != 1:
+            raise ValueError('AllocDiag argument must be a vector!', x)
+        return gof.Apply(self, [x], [x.type.__class__(dtype=x.dtype)()])
+
+    def perform(self, node, inputs, outputs):
+        (x,) = inputs
+        (z,) = outputs
+
+        pdb.set_trace()
+        
+        dim = x.shape[0] + self.offset
+        z = gpuarray.zeros((dim, dim))
```

Comment:
This should be `z[0] = gpuarray.zeros((dim, dim), dtype=node.outputs[0].dtype, context=x.context)`

Also, it might be easier to just do the C code and ignore perform.  You can look at GpuEye which does a similar task.